server/lib/matchmaking/matchmaking-service.ts

```diff
@@ -462,82 +490,49 @@ export class MatchmakingApi {
     }
   }
 
-  private handleLeave = (client: ClientSocketsGroup) => {
-    // NOTE(2Pac): Client can leave, i.e. disconnect, during the queueing process, during the
-    // loading process, or even during the game process.
-    const entry = this.queueEntries.get(client.name)
-    // Means the client disconnected during the queueing process
-    if (entry) {
-      this.queueEntries = this.queueEntries.delete(client.name)
-      this.matchmakers.get(entry.type)!.removeFromQueue(entry.username)
-      this.acceptor.registerDisconnect(client)
-    }
-
-    // Means the client disconnected during the loading process
-    if (this.clientTimers.has(client.name)) {
-      const { mapSelectionTimer, countdownTimer, cancelToken } = this.clientTimers.get(client.name)!
-      if (countdownTimer) {
-        countdownTimer.reject(new Error('Countdown cancelled'))
-      }
-      if (mapSelectionTimer) {
-        mapSelectionTimer.reject(new Error('Map selection cancelled'))
-      }
-
-      cancelToken.cancel()
-
-      this.clientTimers = this.clientTimers.delete(client.name)
-    }
-
-    this.unregisterActivity(client)
-  }
-
   private unregisterActivity(client: ClientSocketsGroup) {
     activityRegistry.unregisterClientForUser(client.userId)
     this.publishToUser(client.name, {
       type: 'status',
       matchmaking: null,
     })
 
-    const user = this.userSockets.getById(client.userId)
-    if (user) {
-      user.unsubscribe(MatchmakingApi.getUserPath(client.name))
+    const userSockets = this.getUserSockets(client.userId)
+    if (userSockets) {
+      userSockets.unsubscribe(MatchmakingService.getUserPath(client.name))
     }
-    client.unsubscribe(MatchmakingApi.getClientPath(client))
+    client.unsubscribe(MatchmakingService.getClientPath(client))
   }
 
-  @Api(
-    '/find',
-    validateBody({
-      type: isValidMatchmakingType,
-      race: validRace,
-    }),
-  )
-  async find(data: Map<string, any>, next: NextFunc) {
-    const { type, race, useAlternateRace, alternateRace, preferredMaps } = data.get('body')
-
-    if (useAlternateRace) {
-      if (!validRace(alternateRace) || alternateRace === 'r') {
-        throw new errors.BadRequest('invalid alternate race')
-      }
-    }
-    if (!Array.isArray(preferredMaps) || preferredMaps.length > 2) {
-      throw new errors.BadRequest('invalid preferred maps, must be an array with length at most 2')
-    }
-
-    const user = this.getUser(data)
-    const client = this.getClient(data)
+  async find(
+    userId: number,
+    clientId: string,
+    type: MatchmakingType,
+    race: RaceChar,
+    useAlternateRace: boolean,
+    alternateRace: AssignedRaceChar,
+    preferredMaps: string[],
+  ) {
+    const userSockets = this.getUserSockets(userId)
+    const clientSockets = this.getClientSockets(userId, clientId)
```

Comment:
This method doesn't do any validation that the client in question should be capable of entering matchmaking (e.g. that the client is an Electron client)

Reply:
Note that I think you could make the argument that `activityRegistry` should be the thing to make that check, but it doesn't do so currently, either, and the error message/code here wouldn't be able to differentiate between "client isn't valid" and "user is already in a game"